dm/dm/config/subtask.go

```diff
@@ -394,10 +395,23 @@ func (c *SubTaskConfig) Adjust(verifyDecryptPassword bool) error {
 		c.MetaSchema = defaultMetaSchema
 	}
 
-	dirSuffix := "." + c.Name
-	if !strings.HasSuffix(c.LoaderConfig.Dir, dirSuffix) { // check to support multiple times calling
-		// if not ends with the task name, we append the task name to the tail
-		c.LoaderConfig.Dir += dirSuffix
+	// adjust dir
+	if c.Mode == ModeAll || c.Mode == ModeFull {
+		isS3, newDir, err := exstorage.AdjustS3Path(c.LoaderConfig.Dir, c.Name+"."+c.SourceID)
+		if err != nil {
+			return terror.ErrConfigLoaderDirInvalid.Delegate(err, c.LoaderConfig.Dir)
+		}
+
+		if isS3 && !c.NeedUseLightning() {
+			return terror.ErrConfigLoaderS3NotSupport.Generate(c.LoaderConfig.Dir)
+		}
+
+		// compatible with other storage and dirSuffix like older
+		dirSuffix := "." + c.Name
+		if !isS3 && !strings.HasSuffix(c.LoaderConfig.Dir, dirSuffix) {
+			newDir = c.LoaderConfig.Dir + dirSuffix
+		}
+		c.LoaderConfig.Dir = newDir
```

Comment:
slightly prefer line 401 to line 414 can be expressed in a clear logic,

```
if isS3 {
  set c.LoaderConfig.Dir
} else {
  check suffix and set c.LoaderConfig.Dir
}
```

Reply:
ok

Reply:
isS3 also need check suffix, because [decode](https://github.com/pingcap/tiflow/blob/master/dm/dm/config/subtask.go#L322) may call `Adjust` again when get subtask config, but i will try to make here better.